Downsample a bam by duplicate sets instead of reads

takutosato opened the pull request:
@lbergelson we've chatted about this about a month ago. I recall you wanted me to write a "ReadWalkerBase," which would be the parent class of ReadWalker and DuplicateSetWalker. I haven't done that just yet—in this code DuplicateSetWalker is a subclass of ReadWalker. I wanted to show you this code first before embarking on further refactoring.

droazen commented:
@lbergelson Are you happy with this branch as-is or do you want further refactoring as @takutosato hinted at above?

takutosato commented:
@droazen or @jamesemery, could one of you take a look at this PR?

lbergelson commented:
I'm reviewing this, sorry for the slow response.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/DownsampleByDuplicateSet.java:
```diff
@@ -0,0 +1,114 @@
+ **/
+public class DownsampleByDuplicateSet extends DuplicateSetWalker {
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc = "")
+    public File outputBam;
```
```suggestion
    public GATKPathSpecifier outputBam;
```

cmnbroad replied:
We're trying to eliminate the use of `File` wherever possible,  since its limiting, and rely the more general `GATKPathSpecifier`.

takutosato replied:
Done

takutosato replied:
@cmnbroad should I avoid createTempFile() in test classes as well? It creates a `File` object.

cmnbroad replied:
@takutosato No - using `createTempFile ` and `File` in tests is fine for now. Thanks.

takutosato replied:
Awesome, thanks!

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/DownsampleByDuplicateSet.java:
```diff
@@ -0,0 +1,114 @@
+    public void onTraversalStart() {
+        super.onTraversalStart();
+        rng = RandomGeneratorFactory.createRandomGenerator(new Random(RANDOM_SEED));
+        outputWriter = createSAMWriter(IOUtils.getPath(outputBam.getAbsolutePath()), false);
```
If you use my `GATKPathSpecifier` suggestion above, you can just call `.toPath()' on the `GATKPathSpecifier` instance here.

takutosato replied:
.toPath -> done.
Hmmm. That's not what I want. But I don't think it sorts the output. Will investigate.

lbergelson replied:
It probably doesn't sort it if the sort order is listed as "unsorted"

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/filters/UMIReadFilter.java:
```diff
@@ -0,0 +1,42 @@
+
+public class UMIReadFilter extends ReadFilter {
+    static final long serialVersionUID = 1L;
+    public static final String UMI_TAG = "RX";
```
Reference `SAMTag.RX.name()` here instead of totally redefining it from scratch.  It's ok to still assign it to a named constant since RX.name() isn't super helpful..

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/filters/UMIReadFilter.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.List;
+
+public class UMIReadFilter extends ReadFilter {
+    static final long serialVersionUID = 1L;
```
`private` is the standard for this for some reason.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/filters/UMIReadFilter.java:
```diff
@@ -0,0 +1,42 @@
+    @Override
+    public boolean test(GATKRead read) {
+        final String readUMI = read.getAttributeAsString(UMI_TAG);
+        final String umi1 = readUMI.split("-", 2)[0];
```
Split once and save the array into a variable.

`Utils.split(String, char)` is substantially faster for non-regex splitting like this.

lbergelson replied:
Could you fix `UMI.getUMI()` while you're at it?  It does the same wasteful thing.    You might consider using that method as well to avoid reimplementing this.

lbergelson replied:
In fact, why not give UMI a string constructor, then make the argument a UMI, and just use UMI.equalsModuloOrder(). It would save reimplementing a bunch of stuff.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/filters/UMIReadFilter.java:
```diff
@@ -0,0 +1,42 @@
+    public static final String UMI_TAG = "RX";
+
+    @Argument(fullName= ReadFilterArgumentDefinitions.UMI_NAME, doc = "umi must have the format 'XXX-XXX'")
+    public String umi;
```
Should we validate that constraint somewhere?  @cmnbroad is there a way to do validation of the arguments in a readfilter?

cmnbroad replied:
Nothing built it. Maybe `UMI` should have a `String` constructor that does the splitting/parsing.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/filters/UMIReadFilter.java:
```diff
@@ -0,0 +1,42 @@
+        final String umi1 = readUMI.split("-", 2)[0];
+        final String umi2 = readUMI.split("-", 2)[1];
+
+        return umi.equals(umi1 + "-" + umi2) || umi.equals(umi2 + "-" + umi1);
```
This would be faster if you reworked it to check `umi.equals(readUmi)` first before deconstructing the string.  In any case, there's no need to rebuild `readUMI`, only the reversed order one.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/filters/UMIReadFilter.java:
```diff
@@ -0,0 +1,42 @@
+        return umi.equals(umi1 + "-" + umi2) || umi.equals(umi2 + "-" + umi1);
+
+//        final File umiFile = new File(umi);
+//        if (umiFile.exists()){
```
dead code.  Do you want to allow many UMIs?

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/UMI.java:
```diff
@@ -0,0 +1,66 @@
+import org.broadinstitute.hellbender.engine.filters.UMIReadFilter;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
+public class UMI {
```
Needs class javadoc.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/UMI.java:
```diff
@@ -0,0 +1,66 @@
+    public final String umi2;
+
+    // Lexicographical ordering of the two (duplex) UMIs;
+    public String umiSmall;
```
final final

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/UMI.java:
```diff
@@ -0,0 +1,66 @@
+
+    public UMI(final GATKRead read){
+        final String readUMI = read.getAttributeAsString(UMIReadFilter.UMI_TAG);
+        this.umi1 = readUMI.split("-", 2)[0];
```
save the array instead of splitting twice, mentioned above

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/UMI.java:
```diff
@@ -0,0 +1,66 @@
+public class UMI {
+    // The umis are stored here in the order they are stored in the read.
+    // e.g. if the read UMI is AAT-CGT, then umi1 = AAT, umi2 = CGT
+    public final String umi1;
```
Should these things have getters?

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/UMI.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    public boolean equalsReadUMI(final GATKRead read){
+        return false;
```
this seems broken

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/UMI.java:
```diff
@@ -0,0 +1,66 @@
+
+    }
+
+    public static Pair<String, String> getUMI(final GATKRead read){
```
Why have this and the constructor?

lbergelson replied:
This is unused.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/UMI.java:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    // Strict match e.g. equalsExactly("AGT-GCT", "GCT-AGT") returns false
+    public boolean equalsExactly(final UMI that) {
```
This should be javadoc.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/UMI.java:
```diff
@@ -0,0 +1,66 @@
+    // Check for whether the two umis came from the same molecule
+    // e.g. equalsModuloOrder("AGT-GCT", "GCT-AGT") returns true
+    //      equalsModuloOrder("AGT-GCT", "AGT-GCT") also returns true
+    public boolean equalsModuloOrder(final UMI that) {
```
This should also be javadoc

lbergelson replied:
Is this post error correction?

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/UMI.java:
```diff
@@ -0,0 +1,66 @@
+        return that.umiSmall.equals(this.umiSmall) && that.umiLarge.equals(this.umiLarge);
+    }
+
+    public String getStandardizedUMI(){
```
Javadoc

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/UMI.java:
```diff
@@ -0,0 +1,66 @@
+
+    public String getStandardizedUMI(){
+        if (umi1.compareTo(umi2) > 0){
+            // umi1 is lexicographically greater e.g. umi1 = TAT, umi2 = AAC
```
This is redundant since you sorted them already, just return umiSmall + umiLong, or better yet just save the combination.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/UMI.java:
```diff
@@ -0,0 +1,66 @@
+import org.apache.commons.lang3.tuple.Pair;
+import org.broadinstitute.hellbender.engine.filters.UMIReadFilter;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
```
This class has lots of unused/untested stuff that probably should be used / tested.  The umi logic should be implemented here and called from the filter.

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,83 @@
+ * http://fulcrumgenomics.github.io/fgbio/tools/latest/GroupReadsByUmi.html
+ */
+public abstract class DuplicateSetWalker extends ReadWalker {
+    DuplicateSet currentDuplicateSet; // TODO: does it make sense to call clear() and recycle the same object?
```
Sounds good. Done

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,83 @@
+ */
+public abstract class DuplicateSetWalker extends ReadWalker {
+    DuplicateSet currentDuplicateSet; // TODO: does it make sense to call clear() and recycle the same object?
+    public static int INITIAL_MOLECULAR_ID = -1;
```
Done

takutosato replied:
removed variable

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/ReadWalker.java:
```diff
@@ -46,7 +46,7 @@ public boolean requiresReads() {
      */
     @Override
-    protected final void onStartup() {
+    protected void onStartup() {
```
Yes I'm going to try. Uno momento.

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,83 @@
+        // write a ReadWalkerBase parent class for both ReadWalker and DuplicateSet Walker, as Louis suggested.
+        super.onStartup();
+        currentDuplicateSet = new DuplicateSet();
+        currentDuplicateSet.setMoleduleId(INITIAL_MOLECULAR_ID);
```
Good points. Done.

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,83 @@
+     * read6: ... MI:Z:1/B ...
+     * read7: ... MI:Z:1/B ...
+     *
+     * Thus it's sufficient to go through the reads in order and collect them in a list until
```
Bam header has `SO:unsorted`.

takutosato replied:
Added a check to ensure that the molecular ID is monotonically increasing and that the tag exists. They should catch it if the input bam is not sorted as required—is that good enough?

lbergelson replied:
Yeah, that sounds great.

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,83 @@
+     *
+     */
+    @Override
+    public void apply(GATKRead read, ReferenceContext referenceContext, FeatureContext featureContext) {
```
done*3

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,83 @@
+        }
+    }
+
+    public abstract void apply(DuplicateSet duplicateSet, ReferenceContext referenceContext, FeatureContext featureContext );
```
done

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,83 @@
+
+    public abstract void apply(DuplicateSet duplicateSet, ReferenceContext referenceContext, FeatureContext featureContext );
+
+    protected boolean rejectDuplicateSet(final DuplicateSet duplicateSet){
```
done*2. The idea is that the subclass can override it this to specify its own criteria for filtering a duplicate set, and I think "reject" is a clearer verb than "filter"

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,83 @@
+
+    protected boolean rejectDuplicateSet(final DuplicateSet duplicateSet){
+        if (!duplicateSet.hasValidInterval()) {
+            logger.info("Duplicate Set with Invalid Intervals");
```
This was a vestige of an earlier bug but I think I can safely remove the logging

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,83 @@
+            }
+        }
+
+        return !duplicateSet.hasValidInterval();
```
rewritten

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+        return SimpleInterval.isValid(contig, fragmentStart, fragmentEnd);
+    }
+
+    public static List<String> getMolecularIDs(final List<GATKRead> reads) {
```
Ok, I understand this now... you need clear javadoc and probably to rename this method since it's very strange that getMolecularID on 2 reads can produce the same result but getMolecularIDs then has 2 values.

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/ReadWalker.java:
```diff
@@ -167,4 +167,16 @@ protected final void onShutdown() {
+            return reads.iterator().next();
+        }
+        // Maybe the correct thing to do is for this to return an error when the user input is empty
+        return null;
```
The method removed

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,83 @@
+            }
+
+            apply(currentDuplicateSet,
+                    new ReferenceContext(reference, currentDuplicateSet.getDuplicateSetInterval()), // Will create an empty ReferenceContext if reference or readInterval == null
```
done

lbergelson commented:
@takutosato Woops, I just realized you might have just wanted a rough look at this and not an in depth review... sorry if this was premature.

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/DownsampleByDuplicateSet.java:
```diff
@@ -0,0 +1,114 @@
+ *
+ * The input bam must have been sorted by UMI using FGBio GroupReadsByUmi (http://fulcrumgenomics.github.io/fgbio/tools/latest/GroupReadsByUmi.html).
+ *
+ * Use this tool to create, for instance, an insilico mixture of duplex-sequenced samples to simulate tumor subclone.
```
done

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/DownsampleByDuplicateSet.java:
```diff
@@ -0,0 +1,114 @@
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME, doc = "")
+    public File outputBam;
+
+    @Argument(fullName = "DS", doc = "This fraction of duplicate sets in the input bam will be retained")
```
done*2

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/DownsampleByDuplicateSet.java:
```diff
@@ -0,0 +1,114 @@
+
+    private static final int RANDOM_SEED = 142;
+    private RandomGenerator rng;
+    private static int numFragments;
```
done

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/DownsampleByDuplicateSet.java:
```diff
@@ -0,0 +1,114 @@
+
+    @Override
+    public void onTraversalStart() {
+        super.onTraversalStart();
```
removed

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/DownsampleByDuplicateSet.java:
```diff
@@ -0,0 +1,114 @@
+    @Override
+    public void onTraversalStart() {
+        super.onTraversalStart();
+        rng = RandomGeneratorFactory.createRandomGenerator(new Random(RANDOM_SEED));
```
No reason, just copied and pasted from another tool. Happy to use Random instead.

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/DownsampleByDuplicateSet.java:
```diff
@@ -0,0 +1,114 @@
+
+    @Override
+    public Object onTraversalSuccess(){
+        outputWriter.close();
```
Done

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/DownsampleByDuplicateSet.java:
```diff
@@ -0,0 +1,114 @@
+
+        // Experiment: only keep duplex
+        if (duplexOnly){
+            final List<String> molecularIDs = DuplicateSet.getMolecularIDs(duplicateSet.getReads());
```
Exactly. Two reads can have the same molecular number, which means they come from the same molecule, but they can come from the different strands. 

I'm going to extract a class for ID, with members ID number (e.g. 1,2,...) and strand (A/B) etc to make this clear

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/DownsampleByDuplicateSet.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    private boolean filterDuplicateSet(final DuplicateSet duplicateSet){
+        if (duplicateSet.getReads().size() % 2 == 1){
```
I believe those are filtered by the upstream fgbio tool

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class DuplicateSet {
```
Renamed to `ReadSetWithSharedUMI` tentatively. Javadoc written.

takutosato commented:
@lbergelson no I appreciate the review. Having read the code, do you think it's worth writing a "ReadWalkerBase," which would be a parent class of ReadWalker and DuplicateSetWalker?

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class DuplicateSet {
```
Done

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+import java.util.stream.Collectors;
+
+public class DuplicateSet {
+    public static final String FGBIO_MOLECULAR_IDENTIFIER_TAG = "MI";
```
Super

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+    public static final String FGBIO_MOLECULAR_IDENTIFIER_TAG = "MI";
+    public static final String FGBIO_MI_TAG_DELIMITER = "/";
+    private int moleculeId = -1; // TODO: extract an ID class.
+    private UMI umi;
```
Removed

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+        reads = new ArrayList<>();
+    }
+
+    public DuplicateSet(final GATKRead read){
```
done

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+        reads.add(read);
+    }
+
+    public void init(GATKRead read){
```
done

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+
+    public void init(GATKRead read){
+        Utils.validate(moleculeId == -1 || moleculeId == getMoleculeID(read),
+                String.format("Inconsisntent molecule IDs: Duplicate set id = %s, read molecule id = %s", moleculeId, getMoleculeID(read)));
```
ooops

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+        paired = false;
+    }
+
+    public List<GATKRead> getReads(){
```
done

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/ReadWalker.java:
```diff
@@ -167,4 +167,16 @@ protected final void onShutdown() {
+        return;
+    }
+
+    public GATKRead peekFirstRead(){
```
done

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+        return getMoleculeID(read) == moleculeId;
+    }
+
+    public void setMoleduleId(GATKRead read){
```
That's true, removed

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+        moleculeId = getMoleculeID(read);
+    }
+
+    public void setMoleduleId(int id){
```
done

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+            return true;
+        }
+
+        if (sameMolecule(read)){
```
Yes, the way the code is written, when this method returns false it the tool exists with error. I'm going to move that check to the beginning of this method to make that clear.

lbergelson commented:
I don't think it's super necessary as long as the weirdness about the double `apply` method is documented.  It wouldn't be bad, but it's a lot of extra work for not too much benefit.  I think we should push it to the future.  If we have more classes like this it might be worth it.  

What *might* be worthwhile is changing the DuplicateSetWalker to be more abstract, so that it is a walker that groups reads by *something* instead of always molecular id.  An obvious similar walker would be one that groups by readname in readname sorted bams, or by cell in single cell RNA bams. Since we don't have an obvious need for either of those though we could also wait until we have a second example of a walker that needs to do this.

tedsharpe commented:
I'm working on a tool that processes aligned contigs from local assemblies that could use the group-by-queryname functionality.  It would be nice not to have to buffer everything that comes through the apply method to process at the end of the traversal.  Not hugely important, but there's your 2nd example.

lbergelson commented:
@tedsharpe @takutosato Lets do it then, but lets do it in a second PR since this one has a ton of comments already?

lbergelson commented:
@tedsharpe Your tool is a little more sophisticated in it's grouping though isn't it?  It's going out and searching for long range mates I thought?  This is just batching things as they come in assuming everything is in order.

lbergelson commented:
]
'''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''

takutosato commented:
@lbergelson making DuplicateSetWalker more abstract in a separate PR sounds good!

tedsharpe commented:
@lbergelson Not in this part.  Just grouping a queryname sorted input by read name.  (Need to gather the primary line together with all supplementary alignments.)  I think we'd find other uses, too.
@takutosato Separate PR is fine with me.

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    public boolean hasValidInterval(){
+        return SimpleInterval.isValid(contig, fragmentStart, fragmentEnd);
```
True. Removed.

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+
+        if (sameMolecule(read)){
+            reads.add(read);
+            if (read.getStart() < fragmentStart){
```
Much cleaner with spanWith. Now checks that the incoming read has the same contig.

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/DownsampleByDuplicateSet.java:
```diff
@@ -0,0 +1,114 @@
+ * This tool address the above problem by dropping a set fraction of _duplicate sets_, rather than reads, at random.
+ * Implicit in this approach is that a read and its mate are dropped or retained together.
+ * While trivial when the input bam is sorted by UMI and query name, this is far from trivial when one attempts
+ * to downsample reads naively with a tool like {@link PrintReads}.
```
Example usage added.

takutosato commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/consensus/DuplicateSet.java:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    /** Returns true if the read was properly added to the duplicate set **/
+    public boolean addRead(final GATKRead read){
```
True, done

takutosato commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/consensus/DownsampleByDuplicateSetTest.java:
```diff
@@ -0,0 +1,129 @@
+                .add("DS", "1.0");
+        runCommandLine(args, DownsampleByDuplicateSet.class.getSimpleName());
+
+        final ReadsDataSource readsDataSource = new ReadsDataSource(Paths.get(out.getAbsolutePath()));
```
done

takutosato commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/consensus/DownsampleByDuplicateSetTest.java:
```diff
@@ -0,0 +1,129 @@
+        final Iterator<GATKRead> iterator = readsDataSource.iterator();
+        while (iterator.hasNext()){
+            final GATKRead read = iterator.next();
+            final String molecularID = read.getAttributeAsString("MI");
```
done

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/ReadWalker.java:
```diff
@@ -99,7 +99,7 @@ public void traverse() {
                     progressMeter.update(readInterval);
                 });
-
+        postProcess();
```
renamed to `finalizeTraversal` and made it final in DuplicateSetWalker.

takutosato commented on src/main/java/org/broadinstitute/hellbender/engine/ReadWalker.java:
```diff
@@ -167,4 +167,16 @@ protected final void onShutdown() {
         super.onShutdown();
     }
+
+    public void postProcess(){
```
done

takutosato commented:
@lbergelson I addressed your comments, cleaned up the code, and wrote tests. I removed UMI.java and UMIReadFilter.java. They will likely come back in a future PR.

droazen commented:
@takutosato Did you want to try to get this into this week's release?

takutosato commented:
@droazen that was my goal but next release is fine too

fleharty commented:
@takutosato Is there anything holding this up so that it can't be ready for the upcoming release?  If it is possible, I would like to see this in the next release.

takutosato commented:
@fleharty nope just pending review.

fleharty commented:
@lbergelson Is there time to re-review this PR before the release, or have we run out of time?

droazen commented:
If @lbergelson is able to do a final review on this branch tomorrow, I'd be happy to wait until it's in before releasing.

takutosato commented:
@droazen incidentally, are you OK with me adding `finalizeTraversal()` do `ReadWalker`? From inside `apply()` we don't know whether the read we're looking at is the last one, so we need to process whatever hasn't been processed before ending traversal.

droazen commented:
@takutosato Why can't you override the existing method `onTraversalSuccess()`?

takutosato commented:
Because the `reference` variable is package private. If I want to call apply() from onTraversalSuccess, it needs to be protected—will push a commit to shortly to show you what I mean.

droazen commented:
@takutosato Just call the protected method `directlyAccessEngineReferenceDataSource()` :)

takutosato commented:
@droazen it looked like it was going to work but then—

```java
@Override
    final protected ReferenceDataSource directlyAccessEngineReferenceDataSource() {
        throw new GATKException("Should never directly access the engine ReferenceDataSource in walker tool classes " +
                "outside of the engine package. Walker tools should get their data via apply() instead.");
    }
```

Also, this is really not the responsibility of the tool class—it should be handled as part of the walker.

tedsharpe commented:
Stash the reference context you get when passing through the original apply method, and use it in onTraversalSuccess when calling the new apply method to push out the final set.

droazen commented:
@tedsharpe The problem with that is that `apply()` gets a `ReferenceContext`, not the `ReferenceDataSource` needed here. I think that an additional finalization method in `DuplicateSetWalker` is probably the cleanest solution for now.

takutosato commented:
@tedsharpe @droazen thanks guys!

Just pushed a commit with suggested changes. I defined `onTraversalSuccess` in `duplicateSetWalker` to process the remaining set of reads. Then the tool subclass calls `super.onTraversalSuccess()` first thing inside its `onTraversalSuccess`.

Is that a good enough workaround for now?

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,116 @@
+    @Argument(fullName = MIN_REQUIRED_READS_NAME, doc = "The mininum total number of reads required in the set", optional = true)
+    private int minimumRequiredReadsPerUMI = DEFAULT_MINIMUM_READS_PER_SET;
+
+    // The user may choose to keep read sets containing both strands (i.d. duplex evidence) only by setting this argument to a positive number
```
Wording.  This would be clearer as "The user may choose to only keep read sets containing both strands by setting this argument to a positive number."  

I misread this at first as the "read sets containing both strands will be kept ONLY if this is set to a positive number"

takutosato replied:
Ah I went back and forth about this and I clearly chose wrong. done.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,116 @@
+    private int minimumRequiredReadsPerUMI = DEFAULT_MINIMUM_READS_PER_SET;
+
+    // The user may choose to keep read sets containing both strands (i.d. duplex evidence) only by setting this argument to a positive number
+    @Argument(fullName = MIN_REQUIRED_READS_PER_STRAND_NAME, doc = "The mininum total number of reads in each strand", optional = true)
```
You can add a minValue = 0 here to keep it positive.

takutosato replied:
done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,116 @@
+    private static final int DEFAULT_MINIMUM_READS_PER_STRAND = 0;
+
+    @Argument(fullName = MIN_REQUIRED_READS_NAME, doc = "The mininum total number of reads required in the set", optional = true)
+    private int minimumRequiredReadsPerUMI = DEFAULT_MINIMUM_READS_PER_SET;
```
I would add a minValue to the argument to make sure it's positive.

takutosato replied:
done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,116 @@
+
+        if (duplicateSetMoleculeNumber < readMoleculeNumber) {
+            // The incoming read's molecule ID does not match that of the current duplicate, meaning we've reached the end of the current set.
+            // Call the apply() method to process the current set and start a new set.
```
this comment should go on the line below, it's not about the rejection stuff

takutosato replied:
done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,116 @@
+            return;
+        }
+
+        // The incoming read has the same UMI as the current set; simply add to the current set
```
I like to put the exclusive conditions in an else block to make it clear that none of the options above fall down into this, but that's a style point that some people disagree with.

takutosato replied:
Yea I didn't like the way the if else blocks looked so I did it this way, hope that's ok.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,116 @@
+            return true;
+        }
+
+        // Check that the read set is paired (this check may not reject some sets that contain unpaired reads)
```
Is there a todo to make this smarter in the future or is % a good enough check?

takutosato replied:
I think this is good enough for now

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/DuplicateSetWalker.java:
```diff
@@ -0,0 +1,130 @@
+    public abstract void apply(ReadsWithSameUMI readsWithSameUMI, ReferenceContext referenceContext, FeatureContext featureContext );
+
+    /**
+     * A child tool class that override onTraversalSuccess() *must* call super.onTraversalSuccess() in order
```
This is a bit inconsistent with the rest of the framework.  Lets consider changing this in a future refactor.

takutosato replied:
Refactor done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/engine/ReadWalker.java:
```diff
@@ -99,7 +99,6 @@ public void traverse() {
                 });
-
         logger.info(countedFilter.getSummaryLine());
     }
```
Some pointless leftover whitespace changes.

takutosato replied:
boy scout rule done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/DownsampleByDuplicateSet.java:
```diff
@@ -0,0 +1,102 @@
+
+    @Override
+    public void onTraversalStart() {
+        rng = new Random(RANDOM_SEED);
```
I would just initialize this where you declare rng and make the variable final.

takutosato replied:
done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/MoleculeID.java:
```diff
@@ -0,0 +1,65 @@
+        return moleculeNumber + ReadsWithSameUMI.FGBIO_MI_TAG_DELIMITER + strand;
+    }
+
+    public static int getMoleculeNumberOfRead(final GATKRead read){
```
These should probably have javadoc too even though they're pretty self explanatory.

takutosato replied:
done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/MoleculeID.java:
```diff
@@ -0,0 +1,65 @@
+     * @returns Counts of reads from each strand, the first element is always larger than the second
+     **/
+    public static Pair<Integer, Integer> countStrands(final List<GATKRead> reads){
+        final int strandACount = (int) reads.stream().filter(r -> getStrandOfRead(r).equals("A")).count();
```
There's definitely a more efficient way of doing this that only loops through once but this is fine.

droazen commented:
@takutosato A separate method called automatically in `DuplicateSetWalker` would be more foolproof than requiring the tool author to remember to call super in `onTraversalSuccess()`, I think.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/ReadsWithSameUMI.java:
```diff
@@ -0,0 +1,106 @@
+    public void addRead(final GATKRead read){
+        Utils.validate(reads.isEmpty() || moleculeID.getMoleculeNumber() == MoleculeID.getMoleculeNumberOfRead(read),
+                String.format("Molecule number of the set and that of the new read don't match: set number = %d, read number = %d", moleculeID.getMoleculeNumber(), MoleculeID.getMoleculeNumberOfRead(read)));
+        Utils.validate(interval.getContig().equals(read.getContig()),
```
There's a `interval.contigsMatch(read)` method that makes this slightly less verbose (and could be implemented slightly more efficiently than getContig.equals(getContig) in some cases.

takutosato replied:
done

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/consensus/ReadsWithSameUMI.java:
```diff
@@ -0,0 +1,106 @@
+    private void init(GATKRead read){
+        Utils.validate(reads.isEmpty(), String.format("Initializing a non-empty set"));
+        moleculeID = new MoleculeID(read);
+        interval = new SimpleInterval(read.getContig(), read.getStart(), read.getEnd());
```
I think you can just write this as `new SimpleInterval(read)`

takutosato replied:
beautiful

takutosato commented:
I think inside `onTraversalSuccess` might be the only place in which we can call such a method automatically—right?

droazen commented:
@takutosato Why not in the `traverse()` method of `DuplicateSetWalker`?

droazen commented:
(ie., override `traverse()` in `DuplicateSetWalker`, call `super.traverse()` (the ReadWalker version) first, then do your extra cleanup work)

takutosato commented:
@droazen that is GENIUS
